Email/confirmable users

Matt-Yorkley opened the pull request:
#### What? Why?

Adds email confirmations to the user model as part of the #946 epic. Supercedes #1690. 

#### What should we test?

User signs up, confirmation email gets sent. 
Clicking the link in confirmation email confirms the user's account. 
User can then log in. 
If the user tries to log in before email is confirmed an appropriate message will show in the login modal.

#### Release notes

Users will now have to confirm their email after sign-up before they can log in.

#### Discourse thread

https://community.openfoodnetwork.org/t/confirmation-emails-registration-flow/237

#### Screenshots

After user signup with login modal:
![screenshot from 2017-08-15 14-10-19](https://user-images.githubusercontent.com/9029026/29317550-58337afc-81c4-11e7-8ea7-cf2d18232e67.png)

Attempting to log in:
![screenshot from 2017-08-15 14-09-59](https://user-images.githubusercontent.com/9029026/29317556-5fa222ac-81c4-11e7-8e7d-f2d714e43f3b.png)

After clicking resend confirmation:
![screenshot from 2017-08-15 14-05-53](https://user-images.githubusercontent.com/9029026/29317560-65eaa1a2-81c4-11e7-9dac-f330de857355.png)

enricostano commented on app/models/spree/user_decorator.rb:
```diff
@@ -18,10 +18,16 @@
+  devise :confirmable, reconfirmable: true
+  handle_asynchronously :send_confirmation_instructions
+  handle_asynchronously :send_on_create_confirmation_instructions
+  # TODO: Later versions of devise have a dedicated after_confirmation callback, so use that
```
Any link to the code of future Devise? Just add it here in the TODO comment please, thanks!

mkllnk commented on app/controllers/user_confirmations_controller.rb:
```diff
@@ -13,10 +13,10 @@ def create
-      set_flash_message(:success, :confirmation_sent) if is_navigational_format?
-    else
-      set_flash_message(:error, :confirmation_not_sent) if is_navigational_format?
+    if successfully_sent?(resource) && is_navigational_format?
```
Looks like this could have been:
```ruby
    if is_navigational_format?
      if successfully_sent?(resource)
        set_flash_message(:success, :confirmation_sent)
      else
        set_flash_message(:error, :confirmation_not_sent)
      end
    end
```

sstead commented:
- [ ] User signs up, confirmation email gets sent.
- [ ] Clicking the link in confirmation email confirms the user's account.
- [ ] User can then log in.
- [ ] If the user tries to log in before email is confirmed an appropriate message will show in the login modal.

Matt-Yorkley commented:
@sstead I've pushed an update with your suggested changes:

- Changed the email template to adjust confirmation message and fix spelling mistake. 
- Changed the login modal to display signup message in the modal instead of reloading and showing in a flash message.
- This also fixes the reload/missing message issue with signup feedback in the /registration page.

I think the functionality for returning the user to the same page they signed up from via the email link will be a separate PR, after this one has gone through.

mkllnk commented:
@Matt-Yorkley Is there any way we could feature-toggle the fact that people can't log in without confirming their email address? We are a bit worried about all the existing users and it would be nice to have a transition period.

sauloperez commented:
> Is there any way we could feature-toggle the fact that people can't log in without confirming their email address?

I would simply run a data migration that confirms all existing users (we assume everyone is confirmed until now, right?) after we run the migration and before we deploy a new release. This way users shouldn't notice any behaviour change and we avoid adding more complexity.

kirstenalarsen commented:
yes please!!

> On 23 Aug 2017, at 6:36 PM, Pau Pérez Fabregat <notifications@github.com> wrote:
> 
> Is there any way we could feature-toggle the fact that people can't log in without confirming their email address?
> 
> I would simply run a data migration that confirms all existing users (we assume everyone is confirmed until now, right?) after we run the migration and before we deploy a new release. This way users shouldn't notice any behaviour change and we avoid adding more complexity.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/openfoodfoundation/openfoodnetwork/pull/1768#issuecomment-324261310>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ACxryNuo-dIwU2YVOTUn5bLdz17Fg5Mxks5sa-SngaJpZM4O3f7G>.
>

Matt-Yorkley commented:
So do we want an on/off toggle, or do we set current users to confirmed?

Matt-Yorkley commented:
I imagine we won't be turning it "off" again, and all new users will have to confirm emails from now on, so maybe @sauloperez's idea is the best option?

enricostano commented:
@Matt-Yorkley yes I would follow @sauloperez 's suggestion. The data migration will do the trick.

daniellemoorhead commented:
>So do we want an on/off toggle, or do we set current users to confirmed?

I say yes to pre-existing users being confirmed. Tis a sensible approach. Let's do it  @Matt-Yorkley :)

mkllnk commented:
@Matt-Yorkley I'm really sorry, but I accidentally overwrote your commits. Can you upload them again? Does anyone else have the commits ab03fc76b13855abc5ad3010cd343ec5b895d983 and 000eba5ee023b147e8b2ffb3d00ae58edee0c82e?

Matt-Yorkley commented:
@mkllnk I've restored those two commits, plus 0289f50 with the adjusted migration which also got deleted.

sstead commented:
- [ ] Changed the email template to adjust confirmation message and fix spelling mistake.
- [ ] Changed the login modal to display signup message in the modal instead of reloading and showing in a flash message. YES
- [ ] This also fixes the reload/missing message issue with signup feedback in the /registration page.

daniellemoorhead commented:
@Matt-Yorkley do you think you'd be able to have this fixed and ready to be staged for @sstead to test by Friday morning AU team (Thursday night your time)?

mkllnk commented:
I can confirm that our default locale `en.yml` is missing some entries which are present in other locales though. It looks like Pierre pre-empted the absence of enterprise confirmation emails. @Matt-Yorkley Is it right that this pull request doesn't make sense without the removal of enterprise confirmations? Have we been to quick thinking we could merge this?

Matt-Yorkley commented:
@mkllnk This initial PR just adds confirmation to the users, yes. I guess there will be a small amount of overlap with user and enterprise confirmations until the PR to remove enterprise confirmations is merged as well. I had hoped we could merge this first, but maybe we'll have to combine them both into a huge PR? The other one (for enterprise confirmations) is already quite big and complex.

mkllnk commented:
@sstead @daniellemoorhead What would you say about user confirmations and enterprise confirmations for a while? Is that okay until we merged the next step? All the existing users don't need to confirm their current account, but they would still need to confirm an enterprise.

@Matt-Yorkley I see how merging this is still appealing. Depending on how the others see the point above, I can imagine to merge it. But I think we should add the missing translations again, otherwise we will end up with a buggy production site.

sstead commented:
@mkllnk I think it's ok to merge this, and have users confirming their enterprises for a while, until the next piece of work is ready. But we should make sure the email is translated correctly before we push it to production.

mkllnk commented:
Just looking at the state of this, I see three outstanding issues:

- merge conflict
- https://github.com/openfoodfoundation/openfoodnetwork/pull/1720/files#r148705378
- https://github.com/openfoodfoundation/openfoodnetwork/pull/1720/files#r148710469

@Matt-Yorkley If you could reply to these, we can let Sally test this soon. It looks like it's almost there. :-)

Matt-Yorkley commented:
This PR has been combined with / merged into #1720.

enricostano commented:
> This PR has been combined with / merged into #1720.

I hope one day we'll stop doing this 😬 🙏 😘
